app/models/miq_queue/put_methods.rb

```diff
@@ -0,0 +1,118 @@
+  def put_or_update(find_options)
+    find_options  = default_get_options(find_options)
+    args_selector = find_options.delete(:args_selector)
+    conds = find_options.dup
+
+    # Since args are a serializable field, remove them and manually dump them
+    #   for proper comparison.  NOTE: hashes may not compare correctly due to
+    #   it's unordered nature.
+    where_scope = if conds.key?(:args)
+                    args = YAML.dump conds.delete(:args)
+                    where(conds).where(['args = ?', args])
+                  else
+                    where(conds)
+                  end
+
+    msg = nil
+    loop do
+      msg = if args_selector
+              where_scope.order("priority, id").detect { |m| args_selector.call(m.args) }
+            else
+              where_scope.order("priority, id").first
+            end
+
+      save_options = block_given? ? yield(msg, find_options) : nil
+      unless save_options.nil?
+        save_options = save_options.dup
+        save_options.delete(:args_selector)
+      end
+
+      # Add a new queue item based on the returned save options, or the find
+      #   options if no save options were given.
+      if msg.nil?
+        put_options = save_options || find_options
+        put_options.delete(:state)
+        msg = put(put_options)
+        break
+      end
+
+      begin
+        # Update the queue item based on the returned save options.
+        unless save_options.nil?
+          if save_options.key?(:msg_timeout) && (msg.msg_timeout > save_options[:msg_timeout])
+            _log.warn("#{format_short_log_msg(msg)} ignoring request to decrease timeout from <#{msg.msg_timeout}> to <#{save_options[:msg_timeout]}>")
+            save_options.delete(:msg_timeout)
+          end
+
+          msg.update_attributes!(save_options)
+          _log.info("#{format_short_log_msg(msg)} updated with following: #{save_options.inspect}")
+          _log.info("#{format_full_log_msg(msg)}, Requeued")
+        end
+        break
+      rescue ActiveRecord::StaleObjectError
+        _log.debug("#{format_short_log_msg(msg)} stale, retrying...")
+      rescue => err
+        raise RuntimeError,
+              _("%{log_message} \"%{error}\" attempting merge next message") % {:log_message => _log.prefix,
+                                                                                :error       => err},
+              err.backtrace
+      end
+    end
+    msg
+  end
+
+  # Find the MiqQueue item with the specified find options, and if not found
+  #   puts a new item on the queue.  If the item was found, it will not be
+  #   changed, and will be yielded to an optional block, generally for logging
+  #   purposes.
+  def put_unless_exists(find_options)
+    put_or_update(find_options) do |msg, item_hash|
+      ret = yield(msg, item_hash) if block_given?
+      # create the record if the original message did not exist, don't change otherwise
+      ret if msg.nil?
+    end
+  end
+
+  def vmdb_env
```

Comment:
Considering a separate PR to move this stuff into it's own module for things like `Vmdb.root`, and `Vmdb.env`, since this is used a lot...